dbms/src/Functions/FunctionsDuration.h

```diff
@@ -88,4 +87,228 @@ class FunctionMyDurationToSec : public IFunction
+        else if (unit == "minute_second")
+            dispatch<ExtractMyDurationImpl::extractMinuteSecond>(col_from, vec_to);
+        else if (unit == "hour_microsecond")
+            dispatch<ExtractMyDurationImpl::extractHourMicrosecond>(col_from, vec_to);
+        else if (unit == "hour_second")
+            dispatch<ExtractMyDurationImpl::extractHourSecond>(col_from, vec_to);
+        else if (unit == "hour_minute")
+            dispatch<ExtractMyDurationImpl::extractHourMinute>(col_from, vec_to);
+        else
+            throw TiFlashException(fmt::format("Function {} does not support '{}' unit", getName(), unit), Errors::Coprocessor::BadRequest);
+
+        block.getByPosition(result).column = std::move(col_to);
+    }
+
+private:
+    using Func = Int64 (*)(Int64);
+
+    template <Func F>
+    static void dispatch(const ColumnPtr col_from, PaddedPODArray<Int64> & vec_to)
+    {
+        if (const auto * from = checkAndGetColumn<ColumnString>(col_from.get()); from)
+        {
+            const auto & data = from->getChars();
+            const auto & offsets = from->getOffsets();
+            if (checkColumnConst<ColumnString>(from))
+            {
+                StringRef string_ref(data.data(), offsets[0] - 1);
+                constantString<F>(string_ref, from->size(), vec_to);
+            }
+            else
+            {
+                vectorString<F>(data, offsets, vec_to);
+            }
+        }
+        else if (const auto * from = checkAndGetColumn<ColumnInt64>(col_from.get()); from)
+        {
+            const auto & data = from->getData();
+            if (checkColumnConst<ColumnInt64>(from))
+            {
+                constantDuration<F>(from->getInt(0), from->size(), vec_to);
+            }
+            else
+            {
+                vectorDuration<F>(data, vec_to);
+            }
+        }
+    }
+
+    template <Func F>
+    static void constantString(const StringRef & from, size_t size, PaddedPODArray<Int64> & vec_to)
+    {
+        vec_to.resize(size);
+        auto from_value = get<Int64>(parseMyDuration(from.toString(), 6));
+        const auto & const_value = F(from_value);
+        for (size_t i = 0; i < size; ++i)
+        {
+            vec_to[i] = const_value;
+        }
+    }
+
+    template <Func F>
+    static void vectorString(
+        const ColumnString::Chars_t & vec_from,
+        const ColumnString::Offsets & offsets_from,
+        PaddedPODArray<Int64> & vec_to)
+    {
+        vec_to.resize(offsets_from.size() + 1);
+        size_t current_offset = 0;
+        for (size_t i = 0; i < offsets_from.size(); i++)
+        {
+            size_t next_offset = offsets_from[i];
+            size_t string_size = next_offset - current_offset - 1;
+            StringRef string_value(&vec_from[current_offset], string_size);
+            auto nano = get<Int64>(parseMyDuration(string_value.toString(), 6));
```

Comment:
better to let `parseMyDuration` support `StringRef` input, so here don't need to copy the string

Reply:
And what happens if the input is DateTime?